pkg/replica/replica.go

```diff
@@ -613,40 +614,37 @@ func (r *Replica) close() error {
 	return r.writeVolumeMetaData(false, r.info.Rebuilding)
 }
 
-func (r *Replica) encodeToFile(obj interface{}, file string) (err error) {
+func (r *Replica) encodeToFile(obj interface{}, file string) (rollbackFunc func() error, err error) {
 	if r.readOnly {
-		return nil
+		return nil, nil
 	}
 
 	tmpFileName := fmt.Sprintf("%s%s", file, tmpFileSuffix)
 
 	defer func() {
-		var rollbackErr error
+		rollbackFunc = func() error {
+			return os.RemoveAll(r.diskPath(tmpFileName))
+		}
+
 		if err != nil {
-			if _, err := os.Stat(r.diskPath(tmpFileName)); err == nil {
-				if err := os.Remove(r.diskPath(tmpFileName)); err != nil {
-					rollbackErr = err
-				}
-			}
+			err = types.GenerateFunctionErrorWithRollback(err, os.RemoveAll(r.diskPath(tmpFileName)))
 		}
-		err = types.GenerateFunctionErrorWithRollback(err, rollbackErr)
 	}()
 
 	f, err := os.Create(r.diskPath(tmpFileName))
 	if err != nil {
-		return err
+		return rollbackFunc, errors.Wrapf(err, "failed to create the tmp file for file %v", r.diskPath(file))
```

Comment:
After returning the error, actually `tmpFileName` has been deleted in the defer function. Knowing you want to accumulate the errors, but why not just use the returned error?

Reply:
1. Yes, the error will be accumulated. It's for helping understand the whole failure chain.
2. The rollback functions are not for accumulating errors. Instead, they allow the upper caller to be atomic: the upper layer either succeeds to execute all functions, or fails in the middle then does rollback for the previous succeeded parts so that everything looks like unchanged.
  1. If this kind of function failed, we will do rollback at the end of (in the defer function) the current function. Then the upper layer is typically no longer need to re-run the rollback function.
  2. In other words, the returned rollback function will be executed (by the upper layer) when the **current function succeeded** but the **subsequent executions failed**.

Reply:
oh I think I missed that purpose. 

The point is `In other words, the returned rollback function will be executed (by the upper layer) when the current function succeeded but the subsequent executions failed.`.